File: 16.3-sum-closest.java
/*
 * @lc app=leetcode id=16 lang=java
 *
 * [16] 3Sum Closest
 */
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        if (nums == null || nums.length < 3) return 0;
        Arrays.sort(nums);
        int closet = nums[0] + nums[1] + nums[2];
        for (int k = nums.length - 1; k >= 0; --k) {
            if (k < nums.length - 1 && nums[k] == nums[k+1]) continue;
            int vk = nums[k];
            int i = 0, j = k - 1;
            while (i < j) {
                int vi = nums[i], vj = nums[j];
                int sum = vi + vj + vk;
                if (d(sum, target) < d(closet, target)) closet = sum;
                if (sum <= target) while (++i < j && nums[i] == vi);
                if (sum >= target) while (i < --j && nums[j] == vj);
            }
        }
        return closet;
    }

    private int d(int a, int b) {
        return a < b ? b - a : a - b;
    }
}

File: go/16.3-sum-closest.go
package main

import (
	"sort"
)

/*
 * @lc app=leetcode id=16 lang=golang
 *
 * [16] 3Sum Closest
 */

// @lc code=start
func threeSumClosest(nums []int, target int) int {
	sort.Sort(sort.IntSlice(nums))
	res := nums[0] + nums[1] + nums[2]
	md := d(res, target)
	for i := 0; i < len(nums); i++ {
		j, k := i+1, len(nums)-1
		for j < k {
			sum := nums[i] + nums[j] + nums[k]
			d := d(sum, target)
			if d < md {
				md, res = d, sum
				if md == 0 {
					return res // = target
				}
			}
			if sum < target {
				j++
			} else {
				k--
			}
		}
	}
	return res
}

func d(sum, target int) int {
	d := sum - target
	if d > 0 {
		return d
	} else {
		return -d
	}
}

// @lc code=end
